pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -603,9 +603,15 @@ public void segmentStoppedConsuming(LLCSegmentName llcSegmentName, String instan
     try {
       HelixHelper.updateIdealState(_helixManager, realtimeTableName, idealState -> {
         assert idealState != null;
-        idealState.getInstanceStateMap(segmentName).put(instanceName, RealtimeSegmentOnlineOfflineStateModel.OFFLINE);
+        Map<String, String> stateMap = idealState.getInstanceStateMap(segmentName);
+        if (stateMap != null) {
```

Comment:
The `stateMap` should never be null. We should not silently return if it is null.

Reply:
Agreed, the statemap can be null only when the segment has somehow been removed from the table (a bug elsewhere). I can log something to that effect.
We should always return "DONE" to the server, otherwise the server may retry the operation, to no effect. So, no point throwing exceptions here, if that is what you mean

Reply:
I decided to throw NPE here if that happens. That is a nasty bug, and we dont want to be checking for nulls everywhere.